Tests/src/InternalTests/CommonTypeTests/Optional.cpp

```diff
@@ -0,0 +1,290 @@
+
+	int ExtraField;
+};
+
+CSP_INTERNAL_TEST(CSPEngine, CommonOptionalTests, OptionalDefaultInitialisationTest)
+{
+	try
+	{
+		Optional<OptionalTestClass> OptionalInstance;
+
+		EXPECT_FALSE(OptionalInstance.HasValue());
+		EXPECT_EQ(&(*OptionalInstance), nullptr);
+	}
+	catch (...)
+	{
+		FAIL();
+	}
+}
+
+CSP_INTERNAL_TEST(CSPEngine, CommonOptionalTests, OptionalNullptrInitialisationTest)
+{
+	try
+	{
+		Optional<OptionalTestClass> OptionalInstance(nullptr);
+
+		EXPECT_FALSE(OptionalInstance.HasValue());
+		EXPECT_EQ(&(*OptionalInstance), nullptr);
+	}
+	catch (...)
+	{
+		FAIL();
+	}
+}
+
+CSP_INTERNAL_TEST(CSPEngine, CommonOptionalTests, OptionalWithPointerInitialisationTest)
+{
+
+	try
+	{
+		auto* TestClassInstancePtr = (OptionalTestClass*) csp::memory::DllAlloc(sizeof(OptionalTestClass));
+		new (TestClassInstancePtr) OptionalTestClass(2);
+
+		Optional<OptionalTestClass> OptionalInstance(TestClassInstancePtr);
+
+		EXPECT_TRUE(OptionalInstance.HasValue());
+		EXPECT_EQ(&(*OptionalInstance), TestClassInstancePtr);
+		EXPECT_EQ((*OptionalInstance).SomeField, TestClassInstancePtr->SomeField);
+		EXPECT_EQ(OptionalInstance->SomeField, TestClassInstancePtr->SomeField);
+	}
+	catch (...)
+	{
+		FAIL();
+	}
+}
+
+CSP_INTERNAL_TEST(CSPEngine, CommonOptionalTests, OptionalWithPointerAndDestructorInitialisationTest)
+{
+	try
+	{
+		bool DestructorRan = false;
+
+		std::function<void(OptionalTestClass*)> CustomDestructor = [&](OptionalTestClass* Pointer)
+		{
+			Pointer->~OptionalTestClass();
+			csp::memory::DllFree(Pointer);
+			DestructorRan = true;
+		};
+
+		{
+			auto* TestClassInstancePtr = (OptionalTestClass*) csp::memory::DllAlloc(sizeof(OptionalTestClass));
+			new (TestClassInstancePtr) OptionalTestClass(2);
+			Optional<OptionalTestClass> OptionalInstance(TestClassInstancePtr, CustomDestructor);
+
+			EXPECT_TRUE(OptionalInstance.HasValue());
+			EXPECT_EQ(&(*OptionalInstance), TestClassInstancePtr);
+			EXPECT_EQ((*OptionalInstance).SomeField, TestClassInstancePtr->SomeField);
+			EXPECT_EQ(OptionalInstance->SomeField, TestClassInstancePtr->SomeField);
+		}
+
+		EXPECT_TRUE(DestructorRan);
+	}
+	catch (...)
+	{
+		FAIL();
+	}
+}
+
+CSP_INTERNAL_TEST(CSPEngine, CommonOptionalTests, OptionalWithExtendedClassPointerInitialisationTest)
+{
+	try
+	{
+		auto* TestClassInstancePtr = (OptionalExtraTestClass*) csp::memory::DllAlloc(sizeof(OptionalExtraTestClass));
+		new (TestClassInstancePtr) OptionalExtraTestClass(3, 4);
+		Optional<OptionalTestClass> OptionalInstance(TestClassInstancePtr);
+
+		EXPECT_TRUE(OptionalInstance.HasValue());
+		EXPECT_EQ((*OptionalInstance).SomeField, TestClassInstancePtr->SomeField);
+		EXPECT_EQ(OptionalInstance->SomeField, TestClassInstancePtr->SomeField);
+
+		csp::memory::DllFree(TestClassInstancePtr);
+	}
+	catch (...)
+	{
+		FAIL();
+	}
+}
+
+CSP_INTERNAL_TEST(CSPEngine, CommonOptionalTests, CopyOptionalWithNoValueInitialisationTest)
```

Comment:
Is there a defined pattern specified somewhere? 

My guess from reading the others is that your suggestion would lead to `OptionalNoValueCopyInitialisationTest` for a test like this. Is that correct?

Reply:
Yeah that sounds about right :)